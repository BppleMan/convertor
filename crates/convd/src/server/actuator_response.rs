use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActuatorResponse<T> {
    pub status: u16,
    pub message: String,
    pub data: Option<T>,
}

impl<T> Default for ActuatorResponse<T>
where
    T: Serialize + DeserializeOwned,
{
    fn default() -> Self {
        Self {
            status: ActuatorResponseStatus::OK,
            message: "OK".to_string(),
            data: None,
        }
    }
}

impl<T> ActuatorResponse<T>
where
    T: Serialize + DeserializeOwned,
{
    pub fn ok() -> ActuatorResponse<T> {
        Self {
            status: 0,
            message: "OK".to_string(),
            data: None,
        }
    }

    pub fn with_data(mut self, data: T) -> Self {
        self.data = Some(data);
        self
    }

    pub fn ok_data(data: T) -> Self {
        Self {
            status: 200,
            message: "OK".to_string(),
            data: Some(data),
        }
    }
}

#[derive(Default, Serialize, Deserialize)]
pub enum ActuatorResponseStatus {
    #[default]
    OK = 0,
}
