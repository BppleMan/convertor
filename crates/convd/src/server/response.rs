mod api_error;
mod api_response;
mod app_error;
mod request_snapshot;

pub use api_error::*;
pub use api_response::*;
pub use app_error::*;
pub use request_snapshot::*;
