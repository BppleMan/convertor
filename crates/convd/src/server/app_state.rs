use crate::server::service::{ClashService, SurgeService};
use convertor::config::ConvertorConfig;
use convertor::config::provider_config::Provider;
use convertor::provider_api::ProviderApi;
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Clone)]
pub struct AppState {
    pub config: Arc<ConvertorConfig>,
    pub api_map: HashMap<Provider, ProviderApi>,
    pub surge_service: SurgeService,
    pub clash_service: ClashService,
}

impl AppState {
    pub fn new(config: ConvertorConfig, api_map: HashMap<Provider, ProviderApi>) -> Self {
        let config = Arc::new(config);
        let surge_service = SurgeService::new(config.clone());
        let clash_service = ClashService::new(config.clone());
        Self {
            config,
            api_map,
            surge_service,
            clash_service,
        }
    }
}
