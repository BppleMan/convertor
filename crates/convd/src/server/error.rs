use axum::http::StatusCode;
use axum::http::header::ToStrError;
use axum::response::{IntoResponse, Response};
use convertor::common::config::proxy_client_config::ProxyClient;
use convertor::url::url_error::{QueryError, UrlBuilderError};
use std::sync::Arc;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum AppError {
    #[error("获取原始非转换配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("没有找到对应的订阅提供者")]
    NoSubProvider,

    // #[error("没有找到对应的代理客户端")]
    // NoProxyClient,
    #[error(transparent)]
    ConvertorUrl(#[from] UrlBuilderError),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    Eyre(#[from] color_eyre::Report),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<AppError>),
    // #[error("读写锁异常: {0}")]
    // RwLockError(String),
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let message = format!("{self:?}");
        let message = console::strip_ansi_codes(&message).to_string();
        (StatusCode::INTERNAL_SERVER_ERROR, message).into_response()
    }
}
