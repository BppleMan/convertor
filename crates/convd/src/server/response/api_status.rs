use crate::server::response::ApiResponse;
use axum::http::{HeaderValue, StatusCode, header};
use axum::response::{IntoResponse, Response};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use tokio_util::bytes::{BufMut, BytesMut};

#[derive(Default, Clone, Serialize, Deserialize)]
pub struct ApiStatus {
    pub code: i64,
    pub message: Cow<'static, str>,
}

impl ApiStatus {
    pub const OK: Self = ApiStatus {
        code: 0,
        message: Cow::Borrowed("ok"),
    };

    pub const ERROR: Self = ApiStatus {
        code: 1,
        message: Cow::Borrowed("error"),
    };
}

impl ApiStatus {
    pub fn with_code(mut self, code: i64) -> Self {
        self.code = code;
        self
    }

    pub fn with_message(mut self, message: impl AsRef<str>) -> Self {
        self.message = Cow::Owned(message.as_ref().to_string());
        self
    }
}

impl<T> IntoResponse for ApiResponse<T>
where
    T: Serialize,
{
    fn into_response(self) -> Response {
        // Extracted into separate fn so it's only compiled once for all T.
        fn make_response(buf: BytesMut, ser_result: serde_json::Result<()>) -> Response {
            match ser_result {
                Ok(()) => (
                    [(
                        header::CONTENT_TYPE,
                        HeaderValue::from_static(mime::APPLICATION_JSON.as_ref()),
                    )],
                    buf.freeze(),
                )
                    .into_response(),
                Err(err) => (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    [(
                        header::CONTENT_TYPE,
                        HeaderValue::from_static(mime::TEXT_PLAIN_UTF_8.as_ref()),
                    )],
                    err.to_string(),
                )
                    .into_response(),
            }
        }

        // Use a small initial capacity of 128 bytes like serde_json::to_vec
        // https://docs.rs/serde_json/1.0.82/src/serde_json/ser.rs.html#2189
        let mut buf = BytesMut::with_capacity(128).writer();
        let res = serde_json::to_writer(&mut buf, &self);
        make_response(buf.into_inner(), res)
    }
}
