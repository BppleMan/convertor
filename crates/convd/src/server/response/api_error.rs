use axum::http::StatusCode;
use axum::http::header::ToStrError;
use axum::response::{IntoResponse, Response};
use convertor::config::proxy_client::ProxyClient;
use convertor::error::{ProviderError, QueryError, UrlBuilderError};
use redis::RedisError;
use std::sync::Arc;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ApiError {
    #[error("获取原始非转换配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("没有找到对应的订阅提供者")]
    NoSubProvider,

    #[error(transparent)]
    UrlBuilderError(#[from] UrlBuilderError),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    ProviderError(#[from] ProviderError),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<ApiError>),

    #[error("Redis 错误: {0:?}")]
    RedisError(#[from] RedisError),

    #[error(transparent)]
    Unknown(#[from] color_eyre::Report),
}

impl IntoResponse for ApiError {
    fn into_response(self) -> Response {
        let message = format!("{self}");
        (StatusCode::INTERNAL_SERVER_ERROR, message).into_response()
    }
}
