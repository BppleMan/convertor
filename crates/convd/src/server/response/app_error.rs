use crate::server::response::{ApiResponse, ApiStatus};
use axum::http::header::ToStrError;
use convertor::config::proxy_client::ProxyClient;
use convertor::error::{ParseError, ProviderError, QueryError, RenderError, UrlBuilderError};
use redis::RedisError;
use std::sync::Arc;
use thiserror::Error;

macro_rules! define_error {
    (
        $(#[$enum_attr:meta])*
        $vis:vis enum $name:ident {
            $(
                $(#[$var_attr:meta])*
                $var:ident
                $( ( $( $(#[$fattr:meta])* $fty:ty ),+ $(,)? ) )?
                $( { $( $(#[$field_attr:meta])* $fname:ident : $ftyt:ty ),+ $(,)? } )?
                ; $vi:expr
                $(,)?
            )*
        }
    ) => {
        $(#[$enum_attr])*
        $vis enum $name {
            $(
                $(#[$var_attr])*
                $var
                $( ( $( $(#[$fattr])* $fty ),+ ) )?
                $( { $( $(#[$field_attr])* $fname : $ftyt ),+ } )?
            ),*
        }

        impl From<$name> for ApiResponse<()> {
            fn from(value: $name) -> Self {
                ApiResponse::error_with_status(value.into())
            }
        }

        impl From<AppError> for ApiStatus {
            fn from(value: AppError) -> Self {
                match value {
                    $($name::$var(e) => Self::new($vi, e),)*
                }
            }
        }

    };
}

define_error! {
    /// 用于表示应用程序中的各种错误类型。
    #[allow(clippy::large_enum_variant)]
    #[derive(Debug, Error)]
    pub enum AppError {
        #[error("获取订阅商原始配置失败, 遇到错误的客户端: {0}")]
        RawProfileUnsupportedClient(ProxyClient) ; 1001,

        #[error("Unauthorized: {0}")]
        Unauthorized(String) ; 1002,

        #[error(transparent)]
        UrlBuilderError(#[from] UrlBuilderError) ; 1003,

        #[error(transparent)]
        QueryError(#[from] QueryError) ; 1004,

        #[error(transparent)]
        ProviderError(#[from] ProviderError) ; 1005,

        #[error(transparent)]
        ParseError(#[from] ParseError) ; 1006,

        #[error(transparent)]
        RenderError(#[from] RenderError) ; 1007,

        #[error(transparent)]
        ToStr(#[from] ToStrError) ; 1008,

        #[error(transparent)]
        Utf8Error(#[from] std::str::Utf8Error) ; 1009,

        #[error(transparent)]
        CacheError(#[from] Arc<AppError>) ; 1010,

        #[error("Redis 错误: {0:?}")]
        RedisError(#[from] RedisError) ; 1011,

        #[error(transparent)]
        JsonError(#[from] serde_json::Error) ; 1012,
    }
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Error)]
pub enum SnapshotError {
    #[error("获取订阅商原始配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    UrlBuilderError(#[from] UrlBuilderError),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    ProviderError(#[from] ProviderError),

    #[error(transparent)]
    ParseError(#[from] ParseError),

    #[error(transparent)]
    RenderError(#[from] RenderError),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<SnapshotError>),

    #[error("Redis 错误: {0:?}")]
    RedisError(#[from] RedisError),

    #[error(transparent)]
    JsonError(#[from] serde_json::Error),
}

// impl From<AppError> for ApiResponse<()> {
//     fn from(value: AppError) -> Self {
//         ApiResponse::error_with_status(value.into())
//     }
// }

// impl From<AppError> for ApiStatus {
//     fn from(value: AppError) -> Self {
//         match value {
//             AppError::RawProfileUnsupportedClient(e) => Self::new(1001, e),
//             AppError::UrlBuilderError(e) => Self::new(1002, e),
//             AppError::Unauthorized(e) => Self::new(1003, e),
//             AppError::QueryError(e) => Self::new(1005, e),
//             AppError::ProviderError(e) => Self::new(1006, e),
//             AppError::ParseError(e) => Self::new(9, e),
//             AppError::RenderError(e) => Self::new(1004, e),
//             AppError::ToStr(e) => Self::new(1007, e),
//             AppError::Utf8Error(e) => Self::new(1008, e),
//             AppError::CacheError(e) => Self::new(1009, e),
//             AppError::RedisError(e) => Self::new(10010, e),
//             AppError::JsonError(e) => Self::new(1011, e),
//         }
//     }
// }
