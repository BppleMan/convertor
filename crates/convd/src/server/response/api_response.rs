use crate::server::response::api_error::ApiError;
use crate::server::response::{AppError, RequestSnapshot};
use axum::http::{HeaderValue, StatusCode, header};
use axum::response::{IntoResponse, Response};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::fmt::Display;
use tokio_util::bytes::{BufMut, BytesMut};

#[derive(Default, Clone, Serialize, Deserialize)]
pub struct ApiResponse<T>
where
    T: serde::Serialize,
{
    pub status: String,
    pub messages: Vec<Cow<'static, str>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request: Option<RequestSnapshot>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
}

impl<T> ApiResponse<T>
where
    T: serde::Serialize,
{
    pub fn ok(data: T) -> Self {
        Self {
            status: "ok".to_string(),
            messages: vec![],
            request: None,
            data: Some(data),
        }
    }

    pub fn with_message(mut self, message: impl Display) -> Self {
        self.messages = vec![Cow::Owned(message.to_string())];
        self
    }

    pub fn with_request(mut self, request: RequestSnapshot) -> Self {
        self.request = Some(request);
        self
    }
}

impl ApiResponse<()> {
    pub fn from_error(status: impl Display, error: impl core::error::Error) -> Self {
        let status = status.to_string();
        let mut messages = vec![Cow::Owned(error.to_string())];
        let mut source = error.source();
        while let Some(src) = source {
            messages.push(Cow::Owned(src.to_string()));
            source = src.source();
        }
        Self {
            status,
            messages,
            request: None,
            data: None::<()>,
        }
    }
}

impl<T> IntoResponse for ApiResponse<T>
where
    T: Serialize,
{
    fn into_response(self) -> Response {
        let mut buf = BytesMut::with_capacity(256).writer();
        match serde_json::to_writer(&mut buf, &self) {
            Ok(()) => (
                [(
                    header::CONTENT_TYPE,
                    HeaderValue::from_static(mime::APPLICATION_JSON.as_ref()),
                )],
                buf.into_inner().freeze(),
            )
                .into_response(),
            Err(err) => {
                let api_error = ApiError {
                    status: StatusCode::INTERNAL_SERVER_ERROR,
                    error: AppError::JsonError(err),
                    request: None,
                };
                api_error.into_response()
            }
        }
    }
}
