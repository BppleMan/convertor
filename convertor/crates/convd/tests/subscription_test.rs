#[path = "./server.rs"]
mod server;

use crate::server::{start_server, ServerContext};
use axum::body::Body;
use axum::extract::Request;
use color_eyre::eyre::{eyre, OptionExt};
use convd::server::response::ApiResponse;
use convertor::common::encrypt::encrypt;
use convertor::config::client_config::ProxyClient;
use convertor::config::provider_config::Provider;
use convertor::testkit::init_test;
use convertor::url::url_builder::HostPort;
use convertor::url::url_result::UrlResult;
use http_body_util::BodyExt;
use tower::ServiceExt;

async fn subscription(
    server_context: &ServerContext,
    client: ProxyClient,
    provider: Provider,
) -> color_eyre::Result<String> {
    let ServerContext { app, app_state, .. } = server_context;
    let sub_url = match app_state.config.providers.get(&provider) {
        Some(provider_config) => provider_config.sub_url.clone(),
        None => return Err(eyre!("未找到提供商配置: [providers.{}]", provider)),
    };

    let secret = app_state.config.secret.clone();
    let enc_secret = encrypt(secret.as_bytes(), secret.as_str())?;
    let enc_sub_url = encrypt(secret.as_bytes(), sub_url.as_str())?;
    let interval = 43200;
    let strict = true;

    let request = Request::builder()
        .uri(format!(
            "/api/subscription/{}/{}?secret={enc_secret}&interval={interval}&strict={strict}&sub_url={enc_sub_url}",
            client.as_ref().to_lowercase(),
            provider.as_ref().to_lowercase()
        ))
        .method("GET")
        .header("host", "127.0.0.1")
        .header("user-agent", concat!("convertor/", env!("CARGO_PKG_VERSION")))
        .body(Body::empty())?;
    let response = app.clone().oneshot(request).await?;

    let actual = String::from_utf8_lossy(&response.into_body().collect().await?.to_bytes())
        .to_string()
        .replace(&enc_secret, &secret)
        .replace(&enc_sub_url, sub_url.as_str())
        .replace(
            sub_url
                .host_port()
                .ok_or_eyre("无法从 sub_url 中提取 host port")?
                .as_str(),
            "mock_host_port",
        );

    Ok(actual)
}

#[tokio::test]
async fn test_subscription_surge_boslife() -> color_eyre::Result<()> {
    init_test();
    let server_context = start_server().await?;
    let url_result = subscription(&server_context, ProxyClient::Surge, Provider::BosLife).await?;
    let url_result: ApiResponse<UrlResult> = serde_json::from_str(&url_result)?;
    insta::assert_json_snapshot!(url_result.data.unwrap());
    Ok(())
}
