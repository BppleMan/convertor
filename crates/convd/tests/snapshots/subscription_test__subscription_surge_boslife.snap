---
source: crates/convd/tests/subscription_test.rs
expression: url_result.data.unwrap()
---
{
  "raw_url": {
    "type": "Raw",
    "server": "http://mock_host_port/subscription?token=bppleman&flag=surge",
    "desc": "订阅商原始订阅配置",
    "path": null,
    "query": null
  },
  "raw_profile_url": {
    "type": "RawProfile",
    "server": "http://127.0.0.1/",
    "desc": "转换前订阅配置",
    "path": "/raw-profile/surge",
    "query": "interval=43200&strict=true&sub_url=http://mock_host_port/subscription?token=bppleman"
  },
  "profile_url": {
    "type": "Profile",
    "server": "http://127.0.0.1/",
    "desc": "转换后订阅配置",
    "path": "/profile/surge",
    "query": "interval=43200&strict=true&sub_url=http://mock_host_port/subscription?token=bppleman"
  },
  "rule_providers_url": [
    {
      "type": {
        "RuleProvider": {
          "name": "DIRECT",
          "option": null,
          "is_subscription": true
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [Subscription]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=DIRECT&policy[is_subscription]=true&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "BosLife",
          "option": null,
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [BosLife]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=BosLife&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "BosLife",
          "option": "no-resolve",
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [BosLife: no-resolve]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=BosLife&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "BosLife",
          "option": "force-remote-dns",
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [BosLife: force-remote-dns]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=BosLife&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "DIRECT",
          "option": null,
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [DIRECT]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=DIRECT&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "DIRECT",
          "option": "no-resolve",
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [DIRECT: no-resolve]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=DIRECT&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    },
    {
      "type": {
        "RuleProvider": {
          "name": "DIRECT",
          "option": "force-remote-dns",
          "is_subscription": false
        }
      },
      "server": "http://127.0.0.1/",
      "desc": "规则集: [DIRECT: force-remote-dns]",
      "path": "/rule-provider/surge",
      "query": "interval=43200&policy[name]=DIRECT&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=http://mock_host_port/subscription?token=bppleman"
    }
  ]
}
